src/main/java/ch/njol/skript/events/EvtGrow.java

```diff
@@ -18,85 +18,216 @@
-	public boolean check(final Event e) {
-		if (evtType == STRUCTURE  && types != null && e instanceof StructureGrowEvent) {
-			return types.check(e, new Checker<StructureType>() {
-				@Override
-				public boolean check(final StructureType t) {
-					return t.is(((StructureGrowEvent) e).getSpecies());
+	public boolean check(Event event) {
+		// Exit early if we need fromTypes, but don't have it
+		if (fromTypes == null && actionRestriction != INTO)
+			// We want true for "on grow:", false for anything else
+			// So check against "OF", which is the first pattern; the one that allows "on grow:"
+			return actionRestriction == OF;
+
+		// Can exit early if we're checking against a structure, but the event isn't a structure grow event
+		// Can also exit early if we're checking against a block, but the event isn't a block grow event AND we're not checking for OF
+		// With OF, we can have `on grow of sapling` or `big mushroom` be a StructureGrowEvent
+		if (eventRestriction == STRUCTURE && !(event instanceof StructureGrowEvent)) {
+			return false;
+		} else if (eventRestriction == BLOCK && !(event instanceof BlockGrowEvent) && actionRestriction != OF) {
+			return false;
+		}
+
+		switch (actionRestriction) {
+			case OF:
+				return checkFrom(event, fromTypes) || checkTo(event, fromTypes);
+			case FROM:
+				return checkFrom(event, fromTypes);
+			case INTO:
+				return checkTo(event, toTypes);
+			case FROM_INTO:
+				return checkFrom(event, fromTypes) && checkTo(event, toTypes);
+			default:
+				assert false;
+				return false;
+		}
+	}
+
+	private static boolean checkFrom(Event event, Literal<Object> types) {
+		// treat and lists as or lists
+		if (types instanceof LiteralList)
+			((LiteralList<Object>) types).setAnd(false);
+		if (event instanceof StructureGrowEvent) {
+			Material sapling = ItemUtils.getTreeSapling(((StructureGrowEvent) event).getSpecies());
+			Skript.info("sapling: " + sapling);
+			return types.check(event, type -> {
+				if (type instanceof ItemType) {
+					return ((ItemType) type).isOfType(sapling);
+				} else if (type instanceof BlockData) {
+					return ((BlockData) type).getMaterial() == sapling;
 				}
+				return false;
 			});
-		} else if (evtType == BLOCK && blocks != null && e instanceof BlockGrowEvent) {
-			assert blocks != null;
-			return blocks.check(e, new Checker<ItemType>() {
-				@Override
-				public boolean check(final ItemType t) {
-					return t.isOfType(((BlockGrowEvent) e).getBlock());
+		} else if (event instanceof BlockGrowEvent) {
+			BlockState oldState = ((BlockGrowEvent) event).getBlock().getState();
+			Skript.info("oldState: " + oldState.getBlockData());
+			return types.check(event, type -> {
+				if (type instanceof ItemType) {
+					return ((ItemType) type).isOfType(oldState);
+				} else if (type instanceof BlockData) {
+					return ((BlockData) type).matches(oldState.getBlockData());
 				}
+				return false;
 			});
 		}
+		Skript.info("event: " + event.getClass());
```

Comment:
Giving the logger a good work-out today aren't we?

Reply:
Haha thanks good catch